### PR TITLE
Trim non cluster topology related changes in the WAGED rebalancer calculation.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1063

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Cluster topology fields are considered as the fundamental attributes of the cluster. For example, instance capacity, resource partition count, partition weight, etc.
If the topology related fields are updated, the WAGED rebalancer will trigger the global rebalance process. Otherwise, the rebalancer will only do local (limited) rebalance to avoid unexpected partition movements.
This change introduces HelixProperty Trimmer to filter out the non-topology related fields so the changes in those fields won't trigger the global rebalancing.

### Tests

- [X] The following tests are written for this issue:

TestHelixPropoertyTimmer

- [x] The following is the result of the "mvn test" command on the appropriate module:

[ERROR] Failures:
[ERROR]   TestRebalanceRunningTask.testFixedTargetTaskAndEnabledRebalanceAndNodeAdded:330 expected:<true> but was:<false>
[ERROR]   TestClusterVerifier.testResourceSubset:225 expected:<false> but was:<true>
[INFO]
[ERROR] Tests run: 1139, Failures: 2, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

Rerun,

[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 53.647 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 59.991 s
[INFO] Finished at: 2020-06-09T14:08:15-07:00
[INFO] ------------------------------------------------------------------------

### Commits

- [X] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [X] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ClusterConfigTrimmer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * The following fields are considered as non-topology related.
+   * HELIX_DISABLE_PIPELINE_TRIGGERS,
+   * PERSIST_BEST_POSSIBLE_ASSIGNMENT,
+   * PERSIST_INTERMEDIATE_ASSIGNMENT,
+   * STATE_TRANSITION_THROTTLE_CONFIGS,
+   * STATE_TRANSITION_CANCELLATION_ENABLED,
+   * MISS_TOP_STATE_DURATION_THRESHOLD,
+   * RESOURCE_PRIORITY_FIELD,
+   * REBALANCE_TIMER_P\ERIOD,
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ClusterConfigTrimmer.java:46]: REBALANCE_TIMER_PERIOD

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ResourceConfigTrimmer.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  private ResourceConfigTrimmer() {
+  }
+
+  @Override
+  protected Map<FieldType, Set<String>> getNonTrimmableFields(ResourceConfig resourceConfig) {
+    Map<FieldType, Set<String>> nonTrimmableFields = new HashMap<>(STATIC_NON_TRIMMABLE_FIELD_MAP);
+    // Also don't trim the application defined preference list in the list fields.
+    // They are fixed and considered as part of the cluster topology.
+    nonTrimmableFields
+        .put(FieldType.LIST_FIELD, resourceConfig.getRecord().getListFields().keySet());
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ResourceConfigTrimmer.java:69]: MapField is not trimmable whatever for user defined the mapping or TaskConfigs.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ClusterConfigTrimmer.java (hunk 1 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ClusterConfigTrimmer.java:46]: Good catch!

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ResourceConfigTrimmer.java (hunk 2 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/ResourceConfigTrimmer.java:69]: Discussed with Junkai offline. The other map fields are for the jobs only. We don't have that usage in the scope of WAGED rebalance for now. And they are not strictly considered as the topology change.

**Comment by jiajunwang** [comment]
This PR is ready to be merged, approved by @dasahcc

**Pull request merged**